### PREDICTING  REMAINING USEFULL LIFE OF JET ENGINES

In [1]:
 # import dependancies
import pandas as pd 

In [ ]:
 # setting up columns  
column_names = ['engine_id', 'cycle'] + \
               [f'operational_setting_{i}' for i in range(1, 4)] + \
               [f'sensor_measurement_{i}' for i in range(1, 22)]

 # Loading the data
train_df = pd.read_csv('train_FD001.txt', sep='\s+', header=None, names=column_names)

In [ ]:
display(train_df.head(5))
print(train_df.shape)
train_df.info()
display(train_df.describe().T)

engine_id  cycle  operational_setting_1  operational_setting_2  \
0          1      1                -0.0007                -0.0004   
1          1      2                 0.0019                -0.0003   
2          1      3                -0.0043                 0.0003   
3          1      4                 0.0007                 0.0000   
4          1      5                -0.0019                -0.0002   

   operational_setting_3  sensor_measurement_1  sensor_measurement_2  \
0                  100.0                518.67                641.82   
1                  100.0                518.67                642.15   
2                  100.0                518.67                642.35   
3                  100.0                518.67                642.35   
4                  100.0                518.67                642.37   

   sensor_measurement_3  sensor_measurement_4  sensor_measurement_5  ...  \
0               1589.70               1400.60                 14.62  ...   
1               1591.82               1403.14                 14.62  ...   
2               1587.99               1404.20                 14.62  ...   
3               1582.79               1401.87                 14.62  ...   
4               1582.85               1406.22                 14.62  ...   

   sensor_measurement_12  sensor_measurement_13  sensor_measurement_14  \
0                 521.66                2388.02                8138.62   
1                 522.28                2388.07                8131.49   
2                 522.42                2388.03                8133.23   
3                 522.86                2388.08                8133.83   
4                 522.19                2388.04                8133.80   

   sensor_measurement_15  sensor_measurement_16  sensor_measurement_17  \
0                 8.4195                   0.03                    392   
1                 8.4318                   0.03                    392   
2                 8.4178                   0.03                    390   
3                 8.3682                   0.03                    392   
4                 8.4294                   0.03                    393   

   sensor_measurement_18  sensor_measurement_19  sensor_measurement_20  \
0                   2388                  100.0                  39.06   
1                   2388                  100.0                  39.00   
2                   2388                  100.0                  38.95   
3                   2388                  100.0                  38.88   
4                   2388                  100.0                  38.90   

   sensor_measurement_21  
0                23.4190  
1                23.4236  
2                23.3442  
3                23.3739  
4                23.4044  

[5 rows x 26 columns]

(20631, 26)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20631 entries, 0 to 20630
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   engine_id              20631 non-null  int64  
 1   cycle                  20631 non-null  int64  
 2   operational_setting_1  20631 non-null  float64
 3   operational_setting_2  20631 non-null  float64
 4   operational_setting_3  20631 non-null  float64
 5   sensor_measurement_1   20631 non-null  float64
 6   sensor_measurement_2   20631 non-null  float64
 7   sensor_measurement_3   20631 non-null  float64
 8   sensor_measurement_4   20631 non-null  float64
 9   sensor_measurement_5   20631 non-null  float64
 10  sensor_measurement_6   20631 non-null  float64
 11  sensor_measurement_7   20631 non-null  float64
 12  sensor_measurement_8   20631 non-null  float64
 13  sensor_measurement_9   20631 non-null  float64
 14  sensor_measurement_10  20631 non-null  flo

count         mean           std        min  \
engine_id              20631.0    51.506568  2.922763e+01     1.0000   
cycle                  20631.0   108.807862  6.888099e+01     1.0000   
operational_setting_1  20631.0    -0.000009  2.187313e-03    -0.0087   
operational_setting_2  20631.0     0.000002  2.930621e-04    -0.0006   
operational_setting_3  20631.0   100.000000  0.000000e+00   100.0000   
sensor_measurement_1   20631.0   518.670000  0.000000e+00   518.6700   
sensor_measurement_2   20631.0   642.680934  5.000533e-01   641.2100   
sensor_measurement_3   20631.0  1590.523119  6.131150e+00  1571.0400   
sensor_measurement_4   20631.0  1408.933782  9.000605e+00  1382.2500   
sensor_measurement_5   20631.0    14.620000  1.776400e-15    14.6200   
sensor_measurement_6   20631.0    21.609803  1.388985e-03    21.6000   
sensor_measurement_7   20631.0   553.367711  8.850923e-01   549.8500   
sensor_measurement_8   20631.0  2388.096652  7.098548e-02  2387.9000   
sensor_measurement_9   20631.0  9065.242941  2.208288e+01  9021.7300   
sensor_measurement_10  20631.0     1.300000  0.000000e+00     1.3000   
sensor_measurement_11  20631.0    47.541168  2.670874e-01    46.8500   
sensor_measurement_12  20631.0   521.413470  7.375534e-01   518.6900   
sensor_measurement_13  20631.0  2388.096152  7.191892e-02  2387.8800   
sensor_measurement_14  20631.0  8143.752722  1.907618e+01  8099.9400   
sensor_measurement_15  20631.0     8.442146  3.750504e-02     8.3249   
sensor_measurement_16  20631.0     0.030000  1.387812e-17     0.0300   
sensor_measurement_17  20631.0   393.210654  1.548763e+00   388.0000   
sensor_measurement_18  20631.0  2388.000000  0.000000e+00  2388.0000   
sensor_measurement_19  20631.0   100.000000  0.000000e+00   100.0000   
sensor_measurement_20  20631.0    38.816271  1.807464e-01    38.1400   
sensor_measurement_21  20631.0    23.289705  1.082509e-01    22.8942   

                             25%        50%        75%        max  
engine_id                26.0000    52.0000    77.0000   100.0000  
cycle                    52.0000   104.0000   156.0000   362.0000  
operational_setting_1    -0.0015     0.0000     0.0015     0.0087  
operational_setting_2    -0.0002     0.0000     0.0003     0.0006  
operational_setting_3   100.0000   100.0000   100.0000   100.0000  
sensor_measurement_1    518.6700   518.6700   518.6700   518.6700  
sensor_measurement_2    642.3250   642.6400   643.0000   644.5300  
sensor_measurement_3   1586.2600  1590.1000  1594.3800  1616.9100  
sensor_measurement_4   1402.3600  1408.0400  1414.5550  1441.4900  
sensor_measurement_5     14.6200    14.6200    14.6200    14.6200  
sensor_measurement_6     21.6100    21.6100    21.6100    21.6100  
sensor_measurement_7    552.8100   553.4400   554.0100   556.0600  
sensor_measurement_8   2388.0500  2388.0900  2388.1400  2388.5600  
sensor_measurement_9   9053.1000  9060.6600  9069.4200  9244.5900  
sensor_measurement_10     1.3000     1.3000     1.3000     1.3000  
sensor_measurement_11    47.3500    47.5100    47.7000    48.5300  
sensor_measurement_12   520.9600   521.4800   521.9500   523.3800  
sensor_measurement_13  2388.0400  2388.0900  2388.1400  2388.5600  
sensor_measurement_14  8133.2450  8140.5400  8148.3100  8293.7200  
sensor_measurement_15     8.4149     8.4389     8.4656     8.5848  
sensor_measurement_16     0.0300     0.0300     0.0300     0.0300  
sensor_measurement_17   392.0000   393.0000   394.0000   400.0000  
sensor_measurement_18  2388.0000  2388.0000  2388.0000  2388.0000  
sensor_measurement_19   100.0000   100.0000   100.0000   100.0000  
sensor_measurement_20    38.7000    38.8300    38.9500    39.4300  
sensor_measurement_21    23.2218    23.2979    23.3668    23.6184

From the output we can see that the dataframe contains 20,621 rows and 26 columns , two of the columns Engine Id and the cycle are integers while the other columns are floats.

In [13]:
 # Getting the number of unique values 
display(train_df.nunique())

engine_id                 100
cycle                     362
operational_setting_1     158
operational_setting_2      13
operational_setting_3       1
sensor_measurement_1        1
sensor_measurement_2      310
sensor_measurement_3     3012
sensor_measurement_4     4051
sensor_measurement_5        1
sensor_measurement_6        2
sensor_measurement_7      513
sensor_measurement_8       53
sensor_measurement_9     6403
sensor_measurement_10       1
sensor_measurement_11     159
sensor_measurement_12     427
sensor_measurement_13      56
sensor_measurement_14    6078
sensor_measurement_15    1918
sensor_measurement_16       1
sensor_measurement_17      13
sensor_measurement_18       1
sensor_measurement_19       1
sensor_measurement_20     120
sensor_measurement_21    4745
dtype: int64

In [18]:
# getting the null values 
display(train_df.isnull().sum())

# Getting duplicated values
display(train_df.duplicated().sum())

engine_id                0
cycle                    0
operational_setting_1    0
operational_setting_2    0
operational_setting_3    0
sensor_measurement_1     0
sensor_measurement_2     0
sensor_measurement_3     0
sensor_measurement_4     0
sensor_measurement_5     0
sensor_measurement_6     0
sensor_measurement_7     0
sensor_measurement_8     0
sensor_measurement_9     0
sensor_measurement_10    0
sensor_measurement_11    0
sensor_measurement_12    0
sensor_measurement_13    0
sensor_measurement_14    0
sensor_measurement_15    0
sensor_measurement_16    0
sensor_measurement_17    0
sensor_measurement_18    0
sensor_measurement_19    0
sensor_measurement_20    0
sensor_measurement_21    0
dtype: int64

np.int64(0)

From the output we can see that there are no duplicated values nor are there any null values in.we will then check the number of cycles each engine goes through untill failure 

In [33]:
display((train_df.groupby('engine_id')['cycle'].max()).sort_values(ascending=False))


engine_id
69    362
92    341
96    336
67    313
83    293
     ... 
58    147
70    137
57    137
91    135
39    128
Name: cycle, Length: 100, dtype: int64

From the output we can see that engine 69 has the highest number of cycles at 362 while egine 39 has the lowest number of cycles at 128.